File: ruby/cell.rb
require 'distances'

class Cell
    attr_reader :row, :column
    attr_accessor :north, :south, :west, :east

    def initialize(row, column)
        @row, @column = row, column
        @links = {}
    end

    def link(cell, bidi=true)
        @links[cell] = true
        cell.link(self, false) if bidi
        self
    end

    def unlink(cell, bidi=true)
        @links.delete(cell)
        cell.unlink(self,false) if bidi
        self
    end

    def links
        @links.keys
    end

    def linked?(cell)
        @links.key?(cell)
    end

    def neighbors
        list=[]
        list << north if north
        list << west if west
        list << south if south
        list << east if east
        list
    end

    def distances
        distances = Distances.new(self)
        frontier = [ self ]

        while frontier.any?
            new_frontier = []

            frontier.each do |frontier_cell|
                frontier_cell.links.each do |linked_cell|
                    next if distances[linked_cell]
                    distances[linked_cell] = distances[frontier_cell] + 1
                    new_frontier << linked_cell
                end 
            end

            frontier = new_frontier
        end

        distances
    end
end

File: c#/cell.cs
namespace Mazes
{
    using System.Linq;
    using System.Collections.Generic;
    public class Cell
    {
        private readonly Dictionary<Cell,bool> links = new Dictionary<Cell,bool>();

        public uint Row { get; }
        public uint Column { get; }
        public Cell North { get; set; }
        public Cell South { get; set; }
        public Cell East { get; set; }
        public Cell West { get; set; }

        public Cell(uint row, uint column) { Row = row; Column = column;}

        public Cell Link(Cell linkingCell,bool bidi = true)
        {
            links[linkingCell]=true;
            if(bidi) linkingCell.Link(this,false);
            return this;
        }

        public Cell Unlink(Cell unlinkingCell,bool bidi=true)
        {
            links.Remove(unlinkingCell);
            if(bidi) unlinkingCell.Unlink(this,false);
            return this;
        }

        public List<Cell> GetAllLinks
        {
            get { return links.Keys.ToList(); }
        }

        public bool IsLinked(Cell neighbourCells)
        {
            try { return links[neighbourCells]; }
            catch { return false; }
        }

        public List<Cell> Neighbours
        {
            get
            {
                List<Cell> returnList = new List<Cell>();
                if(North!=null) returnList.Add(North);
                if(South!=null) returnList.Add(South);
                if(East!=null) returnList.Add(East);
                if(West!=null) returnList.Add(West);
                return returnList;
            }
        }

        public Distances GetDistances()
        {
            Distances distances = new Distances(this);
            List<Cell> frontier = new List<Cell>() {this};
            List<Cell> newFrontier = new List<Cell>();

            while(frontier.Count > 0)
            {               
                newFrontier = new List<Cell>(); 
                foreach(Cell frontierCell in frontier)
                {              
                    foreach(Cell linkedCell in frontierCell.GetAllLinks)
                    {
                        if(distances[linkedCell] == null) 
                        {
                            distances[linkedCell] = distances[frontierCell] + 1;
                            newFrontier.Add(linkedCell);
                        }
                    }
                }

                frontier = newFrontier;
            }

            return distances;
        }
    }
}